### PR TITLE
[primTorch] Add repeat and unfold_copy references

### PR DESCRIPTION
**Pull request by rdspring1**
Add References:

- repeat
- unfold
- expand_as

### CODE REVIEW

**Comment by soulitzer** [comment]
Hey @rdspring1 do you also plan to implement the ref for `repeat_interleave`? I started looking into it, but realized later that you already picked up the packet.

**Comment by rdspring1** [comment]
I've looked into `repeat_interleave`, but it isn't a high priority reference for nvfuser. I wasn't planning on implementing it in the near-term.

File: torch/_prims_common/__init__.py [hunk 1]
```diff
@@ -443,7 +443,8 @@ def validate_exclusive_idx(rank: int, ex_idx: int):
 # "Wraps" a dim (up to one time) for the given rank, allowing
 # dims to be specified using negative indices
 def canonicalize_dim(rank: int, idx: int) -> int:
-    # TODO: add a comment for why this is
+    # For scalar tensors with rank 0, then idx must be in the range [-1, 0].
+    # Otherwise, idx should be in the range [-rank, rank-1].
```
Comment (IvanYashchuk) [inline torch/_prims_common/__init__.py:447]: Thanks for expanding this!

File: torch/_refs/__init__.py [hunk 2]
```diff
@@ -2392,6 +2399,67 @@ def permute(a: TensorLikeType, dims: DimsSequenceType) -> TensorLikeType:
     return prims.transpose(a, _permutation)
 
 
+@register_decomposition(torch.ops.aten.repeat)
+def repeat(a: Tensor, *repeat_shape) -> Tensor:
+    # NOTE: cannot use utils.extract_shape_from_varargs here
+    # because that also validates the shape, but the shape
+    # given to repeat may be "invalid"
+    if len(repeat_shape) == 1 and isinstance(repeat_shape[0], Sequence):
+        repeat_shape = tuple(repeat_shape[0])
```
Comment (IvanYashchuk) [inline torch/_refs/__init__.py:2408]: I know it's taken from the expand function. Let's modify `torch._prims_common.extract_shape_from_varargs` to have optional argument `validate_shape=True` and use it here.

File: torch/_refs/__init__.py [hunk 3]
```diff
@@ -2392,6 +2399,67 @@ def permute(a: TensorLikeType, dims: DimsSequenceType) -> TensorLikeType:
...
+
+    utils.check(
+        len(repeat_shape) >= len(a.shape),
+        lambda: "Number of dimensions of repeat dims can not be smaller than number of dimensions of tensor",
```
Comment (IvanYashchuk) [inline torch/_refs/__init__.py:2412]:
```suggestion
        lambda: "repeat: Number of dimensions of repeat dims can not be smaller than number of dimensions of tensor",
```

File: torch/_refs/__init__.py [hunk 4]
```diff
@@ -2392,6 +2399,67 @@ def permute(a: TensorLikeType, dims: DimsSequenceType) -> TensorLikeType:
...
+        for padded_size, repeat_size in zip(padded_shape, repeat_shape)
+    )
+
+    # return an empty tensor if one of the repeat_shape dimensions is zero
+    if 0 in repeat_shape:
+        return torch.empty(
+            target_shape,
+            dtype=a.dtype,
+            device=a.device,
+            requires_grad=a.requires_grad,
```
Comment (IvanYashchuk) [inline torch/_refs/__init__.py:2434]: Not sure whether we should specify the memory format here? Probably unnecessary as there're 0s in `target_shape`
```suggestion
            requires_grad=a.requires_grad,
            memory_format=utils.suggest_memory_format(a),
```

**Comment by IvanYashchuk** [comment]
@ngimel could you please take a look at this PR?

File: torch/_prims_common/__init__.py [hunk 5]
```diff
@@ -449,6 +449,8 @@ def validate_exclusive_idx(rank: int, ex_idx: int):
 # "Wraps" a dim (up to one time) for the given rank, allowing
 # dims to be specified using negative indices
 def canonicalize_dim(rank: int, idx: int, wrap_scalar: bool = True) -> int:
+    # For scalar tensors with rank 0, then idx must be in the range [-1, 0].
```
Comment (mruberry) [inline torch/_prims_common/__init__.py:452]: Nit: seems like this comment could be a note extending the comment above the function's signature

File: torch/_refs/__init__.py [hunk 6]
```diff
@@ -3157,6 +3221,52 @@ def transpose(a: TensorLikeType, dim0: int, dim1: int) -> TensorLikeType:
 swap_axes = transpose
 
 
+def _get_unfold_copy_shape_stride(
```
Comment (mruberry) [inline torch/_refs/__init__.py:3224]: Comment for this function and its intended use -- shouldn't this be defined before `repeat`, too?

**Review by mruberry — APPROVED** [review]
Cool - see small inline comments and this may require a rebase

**Comment by rdspring1** [comment]
@ngimel Could you look at this?

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**